### ISSUE TITLE
Runtime of internal test case increased dramatically after SLP change

### ISSUE THREAD

**Issue by ormris** [llvm/llvm-project#58476]
An internal test case (C++ and LLVM IR attached), takes over 78 min to run after the following commit. Previously, it completed in about 2 min. On current master, the situation has improved substantially, but is still dramatically slower at just over 15 min. I noticed that there's been performance-related discussions on the original review, so I hope the attached test case will be useful for further investigation. It isn't small, but its structure is regular: [test.zip](https://github.com/llvm/llvm-project/files/9822830/test.zip)

```
  commit 7ea03f0b4e4ec5d91d48ba2976f5adc299089ffd
  Author: Alexey Bataev <a.bataev@outlook.com>
  Date:   Thu Nov 18 08:08:01 2021 -0800

      [SLP]Improve reductions analysis and emission, part 1.
```
##### After 7ea03f0b:
```console
$ time opt -slp-vectorizer -o out.ll test.ll
real    78m7.682s
user    78m7.299s
sys     0m0.236s
$
```
##### Before 7ea03f0b:
```console
$ time opt -S -slp-vectorizer test.ll -o slp-opt.ll
real    1m42.031s
user    1m41.980s
sys     0m0.048s
$
```

##### Current master:
```console
$ time opt -S -slp-vectorizer test.ll -o slp-opt.ll
real    15m34.115s
user    15m33.978s
sys     0m0.096s
$
```

**Comment by EugeneZelenko** [comment]
@alexey-bataev

**Comment by alexey-bataev** [comment]
Have a fix, my time for the provided test after the fix is 0.312 s. Hope to commit it today or next week as NFC patch.

**Comment by alexey-bataev** [comment]
Must be fixed in da4e0f7ac58711ddeaf86943ae82c1928e26801f

**Comment by ormris** [comment]
Looks good. Thanks so much!
